src/Catel.Core/Catel.Core.Shared/Services/ObjectIdGenerator.cs

```diff
@@ -7,19 +7,32 @@
 
 namespace Catel.Services
 {
+    using System;
     using System.Collections.Generic;
+    using System.Linq;
+
+    using Catel.Threading;
 
     /// <summary>
     /// The ObjectIdGenerator class.
     /// </summary>
     /// <typeparam name="TObjectType">The object type</typeparam>
     /// <typeparam name="TUniqueIdentifier">The unique identifier type</typeparam>
     public abstract class ObjectIdGenerator<TObjectType, TUniqueIdentifier> : IObjectIdGenerator<TObjectType, TUniqueIdentifier>
+        where TObjectType : class
     {
         private static Queue<TUniqueIdentifier> _releasedUniqueIdentifiers;
 
         private static readonly object _syncObj = new object();
 
+        private static SortedDictionary<TUniqueIdentifier, WeakReference<TObjectType>> _allocatedUniqueIdentifierPerInstances = new SortedDictionary<TUniqueIdentifier, WeakReference<TObjectType>>();
```

Comment:
Why a sorted dictionary, and not a WeakConditionalTable?

Reply:
Is WeakConditionalTable  available for all platforms?

Reply:
Yep, we use it in more places (I believe a collection change notification wrapper or something like that).

Reply:
See https://github.com/Catel/Catel/search?utf8=%E2%9C%93&q=conditionalweaktable&type=